lib/Basic/Diagnostic.cpp

```diff
@@ -0,0 +1,137 @@
+#include "Basic/Diagnostic.hpp"
+
+namespace glu {
+
+void DiagnosticManager::addDiagnostic(
+    DiagnosticSeverity severity, SourceLocation loc, llvm::StringRef message
+)
+{
+    _messages.emplace_back(severity, loc, message.str());
+
+    // Set the error flag if this is an error or fatal error
+    _hasErrors = severity >= DiagnosticSeverity::Error;
+
+    // Print the diagnostic immediately for better debugging experience
+    printDiagnostic(llvm::errs(), _messages.back());
+}
+
+void DiagnosticManager::error(SourceLocation loc, llvm::StringRef message)
+{
+    addDiagnostic(DiagnosticSeverity::Error, loc, message);
+}
+
+void DiagnosticManager::warning(SourceLocation loc, llvm::StringRef message)
+{
+    addDiagnostic(DiagnosticSeverity::Warning, loc, message);
+}
+
+void DiagnosticManager::note(SourceLocation loc, llvm::StringRef message)
+{
+    addDiagnostic(DiagnosticSeverity::Note, loc, message);
+}
+
+void DiagnosticManager::fatal(SourceLocation loc, llvm::StringRef message)
+{
+    addDiagnostic(DiagnosticSeverity::Fatal, loc, message);
+}
+
+void DiagnosticManager::printDiagnostic(
+    llvm::raw_ostream &os, Diagnostic const &msg
+) const
+{
+    // Don't print anything if location is invalid
+    if (!msg.getLocation().isValid()) {
+        return;
+    }
+
+    // Format the location information
+    os << _sourceManager.getBufferName(msg.getLocation()) << ":"
+       << _sourceManager.getSpellingLineNumber(msg.getLocation()) << ":"
+       << _sourceManager.getSpellingColumnNumber(msg.getLocation()) << ": ";
+
+    // Format the severity prefix with colors if supported
+    switch (msg.getSeverity()) {
+    case DiagnosticSeverity::Note: os << "note: "; break;
+    case DiagnosticSeverity::Warning: os << "warning: "; break;
+    case DiagnosticSeverity::Error: os << "error: "; break;
+    case DiagnosticSeverity::Fatal: os << "fatal error: "; break;
+    }
+
+    // Print the actual message
+    os << msg.getMessage() << "\n";
+
+    if (_sourceManager.getCharacterData(msg.getLocation()) == nullptr) {
+        return;
+    }
+
+    auto lineStart = _sourceManager.getLineStart(msg.getLocation()).getOffset();
+    auto lineEnd = _sourceManager.getLineEnd(msg.getLocation()).getOffset();
```

Comment:
The getOffset function within source location should not exist, source locations are opaque data structures that can only be used by the source manager. The offset isn’t a real offset, there is no buffer in which you can use the offset to get a string out of it.